jet_tracking/tests/conftest.py

```diff
@@ -148,6 +171,69 @@ def jet_control(injector, questar, parameters, diffract):
                       diffract=diffract)
 
 
+def _instantiate_fake_device(dev_cls, name=None, prefix='_prefix',
+                             **specified_kw):
+    '''Instantiate a FakeDevice, optionally specifying some initializer kwargs
+
+    If unspecified, all initializer keyword arguments will default to
+    the string f"_{argument_name}_".
+
+    All signals on the device (and its subdevices) are initialized to either 0
+    or ''.
+    '''
+    sig = inspect.signature(dev_cls)
+    ignore_kw = {'kind', 'read_attrs', 'configuration_attrs', 'parent',
+                 'args', 'name', 'prefix'}
+    kwargs = {name: specified_kw.get(name, f'_{param.name}_')
+              for name, param in sig.parameters.items()
+              if param.kind != param.VAR_KEYWORD and
+              name not in ignore_kw
+              }
+    kwargs['name'] = (name if name is not None else dev_cls.__name__)
+    kwargs['prefix'] = prefix
+    dev = dev_cls(**kwargs)
+
+    devs = [dev]
+    while devs:
+        sub_dev = devs.pop(0)
+        devs.extend([getattr(sub_dev, name)
+                     for name in sub_dev._sub_devices])
+        for name, cpt in sub_dev._sig_attrs.items():
+            sig = getattr(sub_dev, name)
+            try:
+                if cpt.kwargs.get('string', False):
+                    sig.sim_put('')
+                else:
+                    sig.sim_put(0)
+            except Exception as ex:
+                ...
+
+    return dev
+
+
+@pytest.fixture(scope='function')
+def device_instances(injector, questar, offaxis_camera, parameters,
+                     offaxis_parameters, diffract, devices):
+    ns = types.SimpleNamespace()
+    ns.Control = _instantiate_fake_device(devices.Control)
+    ns.CoolerShaker = _instantiate_fake_device(devices.CoolerShaker)
+    ns.Diffract = _instantiate_fake_device(devices.Diffract)
+    ns.Diffract = diffract
+    ns.FlowIntegrator = _instantiate_fake_device(devices.FlowIntegrator)
+    ns.HPLC = _instantiate_fake_device(devices.HPLC)
+    ns.Injector = injector
+    ns.Offaxis = offaxis_camera
+    ns.OffaxisParams = offaxis_parameters
+    ns.Parameters = parameters
+    ns.PressureController = _instantiate_fake_device(devices.PressureController)
+    ns.Questar = questar
+    ns.Selector = _instantiate_fake_device(devices.Selector)
+    ns.SDS = SDS({})
+    ns.SDS.SDS_devices.extend([ns.Selector, ns.CoolerShaker, ns.HPLC,
+                               ns.PressureController, ns.FlowIntegrator])
+    return ns
```

Comment:
I like this pytest fixture pattern, I might steal it

Reply:
By all means do! 👍